src/req.cpp

```diff
@@ -82,21 +82,20 @@ int zmq::req_t::xsend (msg_t *msg_)
             _request_id++;
 
             //  Copy request id before sending (see issue #1695 for details).
-            uint32_t *request_id_copy =
-              static_cast<uint32_t *> (malloc (sizeof (uint32_t)));
-            zmq_assert (request_id_copy);
+            uint32_t request_id_copy;
 
-            *request_id_copy = _request_id;
+            request_id_copy = _request_id;
 
             msg_t id;
             int rc =
-              id.init_data (request_id_copy, sizeof (uint32_t), free_id, NULL);
+              id.init_data (&request_id_copy, sizeof (uint32_t), free_id, NULL);
```

Comment:
As mentioned in the comment, this should change to:

```
rc = id.init_size (sizeof (uint32_t));
...
memcpy (id.data (), &_request_id, sizeof (uint32_t));
```
request_id_copy can be removed entirely.

That's because init_data takes a pointer to user-supplied data buffer, which is useful for large buffers but actually detriment for small buffers (here it's 4 bytes) and as you have found out it can lead to memory leaks if the message is not closed.
Using init_size with a small message on the other hand means it's entirely on the stack, so there's no dynamic memory allocation involved at all.

Reply:
@methodmissing ping - do you have time to look at this? The current version of the PR breaks a few tests. Thanks!

Reply:
Yes, I'll get back to it soon.

Reply:
@TomMD did you have time to look at this? Otherwise I'm happy to pull your branch and do the required fixup. The commit will still be yours for full credit.

Reply:
@bluca I believe I fixed the leak in the manner you desired but please to verify. I just code-monkey'ed in your lines without investing in understanding the `id.*` functions.

If you're curious: this and a prior PR (which I'll revive and see through) was motivated by an analysis tool, MuseDev.  We could turn it on for libzmq and get comments on PRs which are either worthy of discussion or easily ignored.

Typical Muse results are [summaries of project analysis results not necessarily related to the PR](https://github.com/TomMD/libzmq/pull/73#issuecomment-508338184) or [line comments for bugs potentially introduced by the PR](https://github.com/TomMD/libzmq/pull/9).  I've been happy running this for libzmq without any special configuration while other projects benefit from a few rules (like ignoring dead stores) - it's all configurable depending on developer preferences.

Reply:
Thank you, this looks great, merged. Having more analysis tools is a good idea, what do I need to do? Is it free for oss projects?

Reply:
I have enabled the muse app on the repository, let's see how it works